core/vm/logger.go

```diff
@@ -46,6 +61,115 @@ type EVMLogger interface {
+// alternatives
+//
+// PR2: Need to decide where we put all those new stuff
+type CallError struct {
+	error
+	code CallErrorCode
+}
+
+func CallErrorFromErr(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	return &CallError{
+		error: err,
+		code:  callErrorCodeFromErr(err),
+	}
+}
+
+func (e *CallError) Error() string {
+	return e.error.Error()
+}
+
+func (e *CallError) Unwrap() error {
+	return errors.Unwrap(e.error)
+}
+
+func (e *CallError) Code() CallErrorCode {
+	return e.code
+}
+
+type CallErrorCode int
+
+const (
+	CallErrorCodeUnspecified CallErrorCode = iota
+
+	CallErrorCodeOutOfGas
+	CallErrorCodeCodeStoreOutOfGas
+	CallErrorCodeDepth
+	CallErrorCodeInsufficientBalance
+	CallErrorCodeContractAddressCollision
+	CallErrorCodeExecutionReverted
+	CallErrorCodeMaxInitCodeSizeExceeded
+	CallErrorCodeMaxCodeSizeExceeded
+	CallErrorCodeInvalidJump
+	CallErrorCodeWriteProtection
+	CallErrorCodeReturnDataOutOfBounds
+	CallErrorCodeGasUintOverflow
+	CallErrorCodeInvalidCode
+	CallErrorCodeNonceUintOverflow
+	CallErrorCodeStackUnderflow
+	CallErrorCodeStackOverflow
+	CallErrorCodeInvalidOpCode
+
+	// CallErrorCodeUnknown explicitly marks an error as unknown, this is useful when error is converted
+	// from an actual `error` in which case if the mapping is not known, we can use this value to indicate that.
+	CallErrorCodeUnknown = math.MaxInt - 1
+)
+
+func callErrorCodeFromErr(err error) CallErrorCode {
+	switch {
+	case errors.Is(err, ErrOutOfGas):
+		return CallErrorCodeOutOfGas
+	case errors.Is(err, ErrCodeStoreOutOfGas):
+		return CallErrorCodeCodeStoreOutOfGas
+	case errors.Is(err, ErrDepth):
+		return CallErrorCodeDepth
+	case errors.Is(err, ErrInsufficientBalance):
+		return CallErrorCodeInsufficientBalance
+	case errors.Is(err, ErrContractAddressCollision):
+		return CallErrorCodeContractAddressCollision
+	case errors.Is(err, ErrExecutionReverted):
+		return CallErrorCodeExecutionReverted
+	case errors.Is(err, ErrMaxInitCodeSizeExceeded):
+		return CallErrorCodeMaxInitCodeSizeExceeded
+	case errors.Is(err, ErrMaxCodeSizeExceeded):
+		return CallErrorCodeMaxCodeSizeExceeded
+	case errors.Is(err, ErrInvalidJump):
+		return CallErrorCodeInvalidJump
+	case errors.Is(err, ErrWriteProtection):
+		return CallErrorCodeWriteProtection
+	case errors.Is(err, ErrReturnDataOutOfBounds):
+		return CallErrorCodeReturnDataOutOfBounds
+	case errors.Is(err, ErrGasUintOverflow):
+		return CallErrorCodeGasUintOverflow
+	case errors.Is(err, ErrInvalidCode):
+		return CallErrorCodeInvalidCode
+	case errors.Is(err, ErrNonceUintOverflow):
+		return CallErrorCodeNonceUintOverflow
+
+	default:
+		// Dynamic errors
+		if v := (*ErrStackUnderflow)(nil); errors.As(err, &v) {
+			return CallErrorCodeStackUnderflow
+		}
+
+		if v := (*ErrStackOverflow)(nil); errors.As(err, &v) {
+			return CallErrorCodeStackOverflow
+		}
+
+		if v := (*ErrInvalidOpCode)(nil); errors.As(err, &v) {
+			return CallErrorCodeInvalidOpCode
+		}
+
+		return CallErrorCodeUnknown
```

Comment:
Let's log a warning here. I'm afraid it will be easy to forget to add new errors to this list.